### PR TITLE
Fixes SPI CS compilation error with Zephyr 3.4

### PR DESCRIPTION
**Pull request by PetervdPerk-NXP**
zephyrproject-rtos/zephyr#56576 changes CS and thus broke ubxlib with Zephyr 3.4 and newer.

### CODE REVIEW

**Comment by RobMeades** [comment]
Again!?  It's a dynamic world over in Zephyr land :-).  Thanks for posting this, will pull it into our internal test system for checking and let you know if it needs any tweaks.

File: port/platform/zephyr/src/u_port_spi.c [hunk 1]
```diff
@@ -352,6 +358,31 @@ static int32_t setSpiConfig(int32_t spi, uPortSpiCfg_t *pSpiCfg,
...
             }
         }
+#else
+        // Try to set the CS pin based on what the SPI controller
+        // has set for CS pins
+        errorCode = getSpiCsControl(spi, pDevice->pinSelect,
+                                    pDevice->indexSelect,
+                                    &pSpiCfg->spiConfig.cs);
+        if ((errorCode == (int32_t) U_ERROR_COMMON_NOT_FOUND) &&
+                   (pDevice->pinSelect >= 0)) {
```
Comment (RobMeades) [inline port/platform/zephyr/src/u_port_spi.c:368]: Our style checker requires the second line here to be aligned like so:

```
        if ((errorCode == (int32_t) U_ERROR_COMMON_NOT_FOUND) &&
            (pDevice->pinSelect >= 0)) {
```

**Review by RobMeades — CHANGES_REQUESTED** [review]
Just one minor thing to keep our (AStyle) style checker happy, otherwise all good.  If you could make that change I will merge.

File: port/platform/zephyr/CMakeLists.txt [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
+# we need to be able to include generated header files
+zephyr_include_directories()
+
 get_filename_component(UBXLIB_BASE ../../../ ABSOLUTE)
 set(ENV{UBXLIB_BASE} ${UBXLIB_BASE})
 
 # Add environment variables passed-in via U_FLAGS
 if (DEFINED ENV{U_FLAGS})
     separate_arguments(U_FLAGS NATIVE_COMMAND "$ENV{U_FLAGS}")
-    target_compile_options(app PRIVATE ${U_FLAGS})
```
Comment (RobMeades) [inline port/platform/zephyr/CMakeLists.txt:15]: This is an excellent change, thanks for tidying up a "todo", but there's also a pair of `CMakeList.txt` files which form the  `runner` builds that our test system uses, adding-in all of the test files and the test `main()`, see here:

https://github.com/CogniPilot/ubxlib/blob/zephyr-3.4-compile-fix/port/platform/zephyr/runner/CMakeLists.txt

...and here:

https://github.com/CogniPilot/ubxlib/blob/zephyr-3.4-compile-fix/port/platform/zephyr/runner_linux/CMakeLists.txt

Not sure how these should be modified to compile with this change...?

Comment (RobMeades) [reply port/platform/zephyr/CMakeLists.txt:15]: FYI, I tried just search/replace `app` with `ubxlib` in those `CMakeLists.txt` files but that resulted in a different error:

```
CMake Error at /home/ubxlib/.ubxlibpkg/nrfconnectsdk-v2.3.0/zephyr/cmake/modules/extensions.cmake:418 (add_library):
  No SOURCES given to target: app
Call Stack (most recent call first):
  /home/ubxlib/.ubxlibpkg/nrfconnectsdk-v2.3.0/zephyr/cmake/modules/kernel.cmake:209 (zephyr_library_named)
  /home/ubxlib/.ubxlibpkg/nrfconnectsdk-v2.3.0/zephyr/cmake/modules/zephyr_default.cmake:117 (include)
  /home/ubxlib/.ubxlibpkg/nrfconnectsdk-v2.3.0/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /home/ubxlib/.ubxlibpkg/nrfconnectsdk-v2.3.0/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:13 (find_package)
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Comment (RobMeades) [reply port/platform/zephyr/CMakeLists.txt:15]: Also tried adding a `zephyr_library_named(runner)` line and using that as the target in those `CMakeLists.txt` files but got the same error.

**Review by RobMeades — CHANGES_REQUESTED** [review]
The very last commit, see detailed comments, needs a little more work if it is to fly.  If you want to just go with the first three commits, the only issue with them is that the `SYS_INIT` function signature change throws a `clang` error:

`SYS_INIT(ubxlib_preinit, PRE_KERNEL_1, 0);`

"incompatible function pointer types initializing 'int (*)(const struct device *)' with an expression of type 'int (void)'"

For this I think it is acceptable to suppress the warning, so could you make it:

```
// Different Zephyr versions have slightly different function
// signatures for the init function, this to keep clang happy
// NOLINTNEXTLINE(clang-diagnostic-incompatible-function-pointer-types)
SYS_INIT(ubxlib_preinit, PRE_KERNEL_1, 0);
```

If you wanted to revert the fourth commit and just keep the first three, our testing otherwise passes.

Of course it would be very good if we could make the fourth commit work also, but I understand if that's a step too far, maybe not even possible they way things are currently structured.

**Comment by PetervdPerk-NXP** [comment]
> This is an excellent change, thanks for tidying up a "todo", but there's also a pair of CMakeList.txt files which form the runner builds that our test system uses, adding-in a few test-related files and the test main(), see here:

Any step-by-step guide to run that runner build manually? So I can try to trigger the error myself.

> For this I think it is acceptable to suppress the warning, so could you make it:

I think that would work, we could also figure out which version of Zephyr changes that header and then use `#ifdef` guards. We would like to fix it because CI fails on that warning.

**Comment by RobMeades** [comment]
> Any step-by-step guide to run that runner build manually? So I can try to trigger the error myself.

Should be: https://github.com/u-blox/ubxlib/tree/master/port/platform/zephyr/runner, though I admit I don't build it myself manually very often.

> we could also figure out which version of Zephyr changes that header and then use `#ifdef` guards

I think for this kind of thing, better to shut `clang` up on one specific line than get into conditionals: the compiler/linker will stop you if it's really wrong.

**Comment by chrsag** [comment]
Hi together, any update on this? 
Would it be helpful to only merge the 1st commit, as this will likely fix the addressed issue and put the others in another MR?
Thanks

**Comment by RobMeades** [comment]
Agreed, I will take @PetervdPerk-NXP's changes internally and work up a commit which just resolves the problem in the title.  I will then post it back to this repo as a preview branch, if you could possibly test that @chrsag (I don't want to move our test system forward a Zephyr version just yet), then I can commit it internally and push the fix to `master` here.

Would that be OK?

**Comment by RobMeades** [comment]
I have pushed the preview branch here:

https://github.com/u-blox/ubxlib/tree/preview_fix_zephyr_spi_rmea

The change is slightly different, in that I have tried to keep the logic the same and just point to the different places that the data ends-up/comes-from, depending on the way the structures work in the various Zephyr versions; this is so that we don't get even more confused as to which code block is doing what next time they change it :-).

This passes our internal testing, which is on the version of Zephyr that comes with nrfConnect 2.3 (Zephyr 3.2.99 I think): if one or both of @chrsag or @PetervdPerk-NXP could check that it works with Zephyr 3.4 (and if it doesn't, what I need to fix) then I'll get the [if necessary modified] change pushed to `master` here, and will update this issue to let you know that it's done before deleting the preview branch.

**Comment by RobMeades** [comment]
Many thanks for checking that.  Your working version, though different to mine, seems relatively close, and I've just tested it and our test systems says "yes" except for a code style issue: we run AStyle over the code, as of commit eb737d1bc7f91f50183aa79da60dcf90639c80ce there is a Github Action here which should run if you make a PR with your change but if you want to do it in advance it is just the location of the opening brace  on the `if()` statement here:

image

Since you have a tested working 3.4 version that works with 3.2.99 I would be very happy to accept a PR.

**Comment by chrsag** [comment]
I created a MR https://github.com/u-blox/ubxlib/pull/150

**Pull request closed without merging**